docs/proposals/14-reversed-cluster-vpn.md

```diff
@@ -0,0 +1,276 @@
+
+The two containers `vpn-seed-server` and `vpn-shoot-client` are new containers and are not related to containers in the github.com/gardener/vpn project. We will create a new project github.com/gardener/vpn2 for these containers. With this solution we intend to supersede the containers from the github.com/gardener/vpn project.
+
+A service `vpn-seed-server` of type `ClusterIP` is created for each control plane in its namespace.
+
+The `vpn-shoot-client` pod connects to the correct `vpn-seed-server` service via the SNI passthrough proxy introduced with [SNI Passthrough proxy for kube-apiservers](https://github.com/gardener/gardener/blob/master/docs/proposals/08-shoot-apiserver-via-sni.md) port 8132. 
+
+Shoot OpenVPN clients (`vpn-shoot-client`) connect to the correct OpenVPN Server using the http proxy feature provided by OpenVPN. A configuration is added to the envoy proxy to detect http proxy requests and forward the connection attempt to the correct OpenVPN server.
+
+The `kube-apiserver` to shoot cluster connections are established using the API server proxy feature via an envoy proxy sidecar container of the `vpn-seed-server` container. 
+
+The restriction regarding the `192.168.123.0/24` network range in the current VPN solution still applies to this proposal. No other restrictions are introduced. In the context of this GEP a pull requst has been filed to block usage of that range by shoot clusters.
+
+### Performance and Scalability
+
+We do expect performance and throughput to be slightly lower compared to the existing solution. This is because the OpenVPN server acts as an additional hop and must decrypt and re-encrypt traffic that passes through. As there are no low latency, or high thoughput requirements for this connection we do not assume this to be an issue.
+
+### Availablity and Failure Scenarios
+
+This solution re-uses multiple instances of the envoy component used for the kube-apiserver endpoints. We assume that the availability for kube-apiservers is good enough for the cluster VPN as well. 
+
+The OpenVPN client- and server pods are singleton pods in this approach and therefore are affected by potential failures and during cluster-, and control plane updates. Potential outages are only restricted to single shoot clusters and are comparable to the situation with the existing solution today.
+
+### Feature Gates and Migration Strategy
+
+We have introduced a gardenlet feature gate `ReversedVPN`. If both, APIServerSNI and `ReversedVPN` are enabled the proposed solution is automatically enabled for all shoot clusters hosted by the seed. All existing shoot clusters will automatically be migrated during the next reconciliation. We have also added a shoot annotation `alpha.featuregates.shoot.gardener.cloud/reversed-vpn` which can override the feature gate to enable or disable the solution for individual clusters.
```

Comment:
```suggestion
A gardenlet feature gate `ReversedVPN` will be used to enable or disable the reversed tunnel implementation. If both, `APIServerSNI` and `ReversedVPN` are enabled the proposed solution is automatically enabled for all shoot clusters hosted by the seed. All existing shoot clusters will automatically be migrated during the next reconciliation. We have also added a shoot annotation `alpha.featuregates.shoot.gardener.cloud/reversed-vpn` which can override the feature gate to enable or disable the solution for individual clusters.
```

> We have also added a shoot annotation `alpha.featuregates.shoot.gardener.cloud/reversed-vpn` which can override the feature gate to enable or disable the solution for individual clusters.

I vote to remove this annotation. Or to use it only to disable the reversed tunnel per shoot. Why - because the gardenlet may be running with `APIServerSNI=false` and then a shoot with `alpha.featuregates.shoot.gardener.cloud/reversed-vpn=true` will not be able to use to reversed tunnel.

In general, what will happen with gardenlet started with?
```
APIServerSNI=false
ReversedVPN=true
```
My opinion is that this is invalid configuration as long as the `APIServerSNI` is not removed and gardenlet has to panic on start-up.

Reply:
Fair point but what I/we would like to avoid is having all clusters migrated automatically. I also don't want to enable this feature without a feature gate. 

Yes, I agree that the gardenlet should panic on the following configuration:

```
APIServerSNI=false
ReversedVPN=true
```

I assume `APIServerSNI=true` will become the default and later on `ReversedVPN=true` as well so this is only a temporary panic situation due to inconsistent configuration.

Yes, `alpha.featuregates.shoot.gardener.cloud/reversed-vpn=false` could be used to opt out but shouldn't it be opt-in? I imagine us enabling `ReversedVPN=true` in the core project and something breaks. I'd rather start to enable it one by one.

Reply:
Will push updated description on what has been agreed between @vpnachev, @DockToFuture, and @ScheererJ: gardenlet will panic on inconsistent configuration.